### ISSUE TITLE
Anyway to get the pixelbuffer for the video chat as it's being captured via an extended QBRTCVideoCapture class?

### ISSUE THREAD

**Issue by astanton** [QuickBlox/quickblox-ios-sdk#404]
I'm currently looking at the video chat sample app.  

My use case is that I need to get the video stream from the client before it is sent to the receiver and I want to do some stuff to the pixel buffer before it is being sent over, such as changing it to black/white and things like that.

So far what I have been trying to do is take a look at the screen sharing and how that is done, and trying to write a class that extends `QBRTCVideoCapture` similar to `QBRTCScreeCapture` class in the sample video chat app.

I see that in the `SharingViewController` the `self.session.localMediaStream.videoTrack.videoCapture` is set to an instance of the `QBRTCScreenCapture` class, so that is what I am looking at and trying to do something in there.

I see that the `sendPixelBuffer` method is being called every 12 frames (5 times per 60fps).  And I see that it creates a pixelbuffer that is sent to the videoframe that is sent to the super function

What I tried to do was extend that class by the `AVCaptureVideoDataOutputSampleBufferDelegate` delegate, and I have then set up my own camera capture and I have the `- ( void ) captureOutput (AVCaptureOutput * ) captureOutput didOutputSampleBuffer: ( CMSampleBufferRef ) sampleBuffer fromConnection: ( AVCaptureConnection * ) connection`delegate implemented, and I am creating a pixelBuffer from that of the current image that is being captured.

I have verified that my image is correct by saving it to my phone and I see it's getting a black/white image.

So once I have the pixel buffer, I'm setting it to a member variable that is then being sent into the videoFrame in the `sendPixelBuffer` function.  

Now this is KIND of working, in that I have seen a few frames being streamed over to the other phone, but at a VERY slow/unnaceptable frame rate, and then it crashes down in the `[super sendVideoFrame:videoFrame];` call in a `CVPixelBufferUnlockBaseAddress(self.currentPixelBufferRef, 0)` call.  

So I am wondering if there is an easier way to do this, rather than trying to create my own class and video feed like this, and if I do have to do it this way, what exactly am I doing wrong?  

The frame rate on the default videoCapture object that is used for the normal video chat in the app is VERY smooth and performant, but like I said when I tried it this way, it was VERY choppy for the 4-5 frames that worked (I'm talking like 1FPS) before it crashes.

Any ideas?

**Comment by astanton** [comment]
So quick update...

I got it working somewhat.

What I did was instead of calling `[super sendVideoFrame:videoFrame]` from the method that is called on the `CADisplayLink' interval, I instead call it from the `captureOutput' delegate method instead.

When I first tried this it did not work for me, but then I realized my timeStamp on the video frame was not being set correctly.  So what I'm doing now is just setting a timeStamp variable in the `CADisplayLink` method, and then just pass that to the videoFrame.

It seems to be working alright ... at first.  It appears that after a while, the captureOutput delegate method just starts to get called less frequently, and in turn almost comes to a complete stop.  

Any ideas why that could be happening?

**Comment by VitaliyGuru** [comment]
Hi @astanton,

Could you please enable stats report and send us this information?

`[QBRTCConfig setStatsReportTimeInterval:1.f];` - method for enabling stats report.

In your controller you should adopt QBRTCClientDelegate and use [this delegate method](https://github.com/QuickBlox/quickblox-ios-sdk/blob/master/sample-videochat-webrtc/sample-videochat-webrtc/ViewControllers/CallViewController/CallViewController.m#L353-404).

Also can you set up a test project with this issue being reproducible and attach it here?

**Comment by astanton** [comment]
@VitaliyGuru I've actually narrowed down what the exact issue is that is causing the slowness and I am not sure if it is related to anything QB or not at this point.  I didn't have too much time to debug after I got to that point last night.

So in my `captureOutput` delegate method I am doing face detection on a `CIImage` that is created from the `CVPixelBufferRef`.  

I'm simply using Apple's built in `CIDetector` class and calling the `featuresInImage:` method with the image. 

I currently have this method being called 30 times a second (once each frame for 30fps) and while it is working, after about 3-4 seconds I started to notice it slows down a lot, and then basically freezes up the app and it won't even hit the delegate method anymore.

When I comment out the code to do this face detection, it works as expected and is nearly as performant as the default video chat view.  There is minimal delay just like the default one.  It does seem however after a few seconds that the delay MAY get slightly greater.  But I'm talking like if say the default delay was like 50ms, this delay is 100 or 150ms.  Nothing major at all and what I would call acceptable for video chat.  But I only think I notice this because of how much I'm testing things out.

I can try enabling the stats report later this evening when I am working on this again though and post the results.  I will also see about getting a sample project up for this as well.

But do you have any idea off of the top of your head why using a `CIDetector` and calling that method inside of this delegate method could be causing such slowdown after it performs the detection a bunch of times?  

I actually have this exact face detection code running on our current OpenTok implementation of this, and it works just fine, so that is why I'm a bit confused as to why it's working there but not here since it's the identical code.  We're looking to transition off of OpenTok to QuickBlox however which is why I'm doing this current research and implementation, to be sure we can move our whole app over to it.

Thanks.

**Comment by VitaliyGuru** [comment]
@astanton , Unfortunately at this moment we can't answer to you question without stats and project.
 They may help us to resolve the issue with the face detection.

**Comment by astanton** [comment]
No problem, I will get the stats in there tonight and copy/paste results.  I'll see what I can do about getting a test project up too, thanks.

**Comment by astanton** [comment]
@VitaliyGuru I got a chance to add the stats report tot he code.  I'll attach the file instead of pasting it because it's quite long.  

I don't know what I'm looking at really but I can definitely see the time rate next to the "opus" goes up from low 90 to over 500.  But other than that I'm not sure what I'm looking at.

[report.txt](https://github.com/QuickBlox/quickblox-ios-sdk/files/228840/report.txt)

**Comment by astanton** [comment]
@VitaliyGuru so I have been messing with this all night and while I thought I made some progress, it still clearly has the performance issue where after a few seconds you can see a performance hit.  

I've tried sending the captured frame at 15FPS with face detection done every frame, I've tried doing it at 20FPS with detection every frame, I've tried 30FPS with detection done every 4 frames, and I've done it at 60fps with detection every 4 frames.  All of these clearly have performance issues which are noticeable after a few seconds.

If you comment out the face detection code, it works just as expected and looks great, with no degradation as time goes on.

I've attached the compressed project file that I've been working on here:

[sample-videochat-webrtc.zip](https://github.com/QuickBlox/quickblox-ios-sdk/files/229076/sample-videochat-webrtc.zip)

This is just the sample video app that you all provide, with a few modifications, which are primarily in the `QBRTCScreenCapture` class.  It also uses the front facing camera instead of the one on the rear.

1. When you press the screen share button from the call view, this will take you to the custom video capture (just like it took you to the screen share).

2. When in this mode, it still shows the images on your screen like it's the default screen share view that you can scroll through.  However it is actually streaming your front camera looking right at you, and you can clearly see this on another device when testing this between 2 people.

3. There is an NSLog statement in the console that will say whether a face is detected or not.  You can clearly see from the speed of the logs slowing down, that the performance starts to slow down.

4. It is currently set at 60fps and doing face detection every 4 frames.  If you want to change that, it's in the `didSetToVideoTrack` method (as I'm sure you know), and to change the detection every 4 frames, in the `captureOutput` delegate method, I'm just doing a `self.faceDetectorFrameCounter % 4 == 0` check whether or not to do face detection, so you can just change that or remove that check if you want to see face detection during all frames.

5. Keep in mind that I've been trying out and testing a bunch of stuff, so it might not be the prettiest of code right now, and that I'm also using this basically as a proof of concept before I move our app over to QuickBlox once I can figure this out :)

I believe that is all.  If you have any other questions please ask me and I will get back to you asap. 

Thanks for the help.

**Comment by Pro100Andrey** [comment]
Hi @astanton, please use h264 codec - best performance (hardware acceleration, iOS8+):
```objective-c
+[QBRTCConfig mediaStreamConfiguration].videoCodec = QBRTCVideoCodecH264;
```

**Comment by astanton** [comment]
@Pro100Andrey that is the codec that is already being used.  I verified this by putting a break point in the face detecting code and checking at that time.

It's set in the AppDelegate with the sample project to be that value.

**Comment by Pro100Andrey** [comment]
@astanton from your stats logs - `VS (enc) 512Kbps/1.70Mbps | (sent) 277Kbps/2.69Mbps | 54ms | **VP8**`
must be h264

VP8 don't have hardware acceleration only software

**Comment by VitaliyGuru** [comment]
@astanton , Do you use simulator for testing or device?

**Comment by astanton** [comment]
@Pro100Andrey Hmmm I am not really sure then because it is definitely set to that in the configuration.  And as mentioned I put a break point in the face detection method and checked it in the console at the time.  

@VitaliyGuru I am using an iPhone 6 device on iOS 8.2 as the one that is sharing the screen doing face detection.  The receiver phone is an iPhone 5 device on iOS 9.2.

**Comment by Pro100Andrey** [comment]
@astanton Pease check `[QBRTCConfig mediaStreamConfiguration].videoCodec` before start (side A) and accept (side B) call, must be h264 + enable `[QBRTCConfig setLogLevel:QBRTCLogLevelVerboseWithWebRTC];` and provide full logs from A & B

**Comment by astanton** [comment]
Even when I just use the "normal" video call on the sample application (not using my custom capture class), with `[QBRTCConfig mediaStreamConfiguration].videoCodec = QBRTCVideoCodecH264;` being called in the AppDelegate, the report is reporting `VPS`. 

I just downloaded a clean version of the sample video chat as well, and put the `[QBRTCConfig mediaStreamConfiguration].videoCodec = QBRTCVideoCodecH264;` call in the `CallViewController.m` at the initialization method, then turned on reporting, and it is still reporting VPS.

@Pro100Andrey okay I will give that a try now and report back.

**Comment by Pro100Andrey** [comment]
@astanton Did you use the [sample](https://github.com/QuickBlox/quickblox-ios-sdk/tree/master/sample-videochat-webrtc) from the master branch?

**Comment by Pro100Andrey** [comment]
@astanton we use [QBRTCConfig mediaStreamConfiguration] before start / accept call

**Comment by astanton** [comment]
@Pro100Andrey yes I was using the master one.  I pulled it about 4-5 days ago.

I just put a break point in in the `CallViewController - (void)start` method, right there start and accept are called, and on both devices A and B, before starting/accepting, the codec is set to H264.

`0x1683c1a0> state: Idle
(lldb) po [QBRTCConfig mediaStreamConfiguration].videoCodec
QBRTCVideoCodecH264`

I also could not add the call `[QBRTCConfig setLogLevel:QBRTCLogLevelVerboseWithWebRTC]` as it is saying that the QBRTCLogLevelVerboseWithWebRTC does not exist.  Could this mean that there is an old framework with the sample project or something?

I wasn't sure if you meant to call setLogLevel on QBRTCSettings or not either since that is where I turned it off, but that is the same problem where it can't recognize that variable.

**Comment by Pro100Andrey** [comment]
@astanton Sorry  `[QBRTCConfig setLogLevel:QBRTCLogLevelVerboseWithWebRTC]` this is my development branch (private)

**Comment by astanton** [comment]
@Pro100Andrey no problem.  

I am not setting the `[QBRTCConfig mediaStreamConfiguration]` to anything other than H264, I just searched the code for that call and it only shows up with that option.  

The project is attached a few posts up as well if you wanted to check it out in the actual project.

**Comment by Pro100Andrey** [comment]
@astanton Give me time, i need investigate your sample

**Comment by astanton** [comment]
@Pro100Andrey no problem I appreciate the help.  

I have to run to work shortly so unfortunately I won't be able to respond and try things out until later this evening (it is 7:45am EST where I am at right now).

**Comment by Pro100Andrey** [comment]
@astanton 
1). Replace `QBRTCCameraCapture` with `FaceDetectroCamera` in `CallViewController.m`:
[FaceDetectorCamera.zip](https://github.com/QuickBlox/quickblox-ios-sdk/files/230200/FaceDetectorCamera.zip)
```objective-c
...
@property (strong, nonatomic) FaceDetectorCamera *cameraCapture;
...
- (void)configureCameraCapture {
    
    if (self.session.conferenceType == QBRTCConferenceTypeVideo) {
        
        Settings *settings = [SampleCore settings];
        self.cameraCapture = [[FaceDetectorCamera alloc] initWithVideoFormat:settings.videoFormat
                                                                    position:AVCaptureDevicePositionFront];
        [self.cameraCapture startSession];
    }
}
```

You're  using main queue for face detection (Your code from the sample):
```Objective-c
        // do face detection every 4 frames
        if (self.faceDetectorFrameCounter % 4 == 0) {
            ---------->  dispatch_async( dispatch_get_main_queue(), ^{
                CIImage *ciImage = [[CIImage alloc] initWithCVPixelBuffer:imageBuffer];
                [self doFaceDetection:ciImage];
            });
        }
```
Please check FaceDetectorCamera.m (:695):
```Objective-C
 CVPixelBufferRef pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer);
    
    CVPixelBufferLockBaseAddress(pixelBuffer, kCVPixelBufferLock_ReadOnly);
    
    CFDictionaryRef attachments = CMCopyDictionaryOfAttachments(kCFAllocatorDefault, sampleBuffer, kCMAttachmentMode_ShouldPropagate);
    CIImage *ciImage = [[CIImage alloc] initWithCVPixelBuffer:pixelBuffer options:(NSDictionary *)CFBridgingRelease(attachments)];
    
    if (_faceDetectorFrameCounter == 15) {
        
        ---------> dispatch_async(_faceDetectorQueue, ^{
            
            // do face detection every 4 frames
            NSArray *features = [_faceDetector featuresInImage:ciImage];
            
            if (features.count > 0) {
                
                NSLog(@"Found face");
            }
            else {
                
                NSLog(@"Did not find face");
            }
        });
        
        _faceDetectorFrameCounter = 0;
    }
    
    _faceDetectorFrameCounter ++;
 ```

2). Please use v 2.1.1 QuickbloxWebRTC.framework for h264. v2.2 has problem with this codec, we will fix it in new version

**Comment by astanton** [comment]
@Pro100Andrey Thanks for all of the help I appreciate it.

So was the main issue the version of the framework then?  Or was it also other stuff that was being done wrong?  I see that the file you have attached is a lot different than what I have attached so I am curious if it was also other stuff going on.  The AV capture I had set up was from a tutorial that I found on the net.  I'm very noob to all of this stuff, which is why we're looking for a solution such as QuickBlox :)

As for the face detection on the main thread, that was just me trying various different things that I had left in there, all of which kept giving me the same results when I didn't have it in there.  I was really grasping last night just trying to figure out something, because the fact that it worked fine then steadily started to perform worse and worse had me really curious as to what could be causing that, and it was _only _happening when I was doing the face detection.

As for the older versions of the frameworks, where do I find those?  The versions that I see on github are at version 2.7.2 so I'm not sure how that relates to v2.2 like you are talking about.  

I really REALLY appreciate the help on this and can't wait to get home later and try this out!  Once I get all of this working we're officially switching over to QuickBlox from OpenTok and I will begin integrating this into the app.  You will have a new happy customer!

**Comment by Pro100Andrey** [comment]
@astanton [v2.1.1] (https://github.com/QuickBlox/quickblox-ios-sdk/releases/tag/WebRTCv2.1.1)

**Comment by astanton** [comment]
Ah okay just the WebRTC portion I saw that earlier.

So I installed that one and it didn't seem to make much of a difference performance wise.

BUT ... with the file you linked to that is working 100% as needed as it is!  I see that you basically extended a class at a higher up level so that you have more control over the stream that is being captured it seems.  I kind of get it but also kind of don't.  So I am not sure if I should just use the v2.2 or v2.1.1 since it didn't seem to do anything different performance wise, although I did see in the report that it says the correct codec now.

I really do appreciate all the help, you've now solidified that we can transition over to QuickBlox and I'm going to begin that transition right now.  

We're using Parse for our backend and I saw earlier that you all have a sample app to connect QB to Parse so I'm going to check that out now as well.

Thanks again!
